StyleCop.Analyzers/StyleCop.Analyzers/AnalyzerExtensions.cs

```diff
@@ -23,72 +24,19 @@ internal static class AnalyzerExtensions
         /// </summary>
         /// <param name="context">The analysis context.</param>
         /// <param name="action">Action to be executed at completion of parsing of a document.</param>
-        public static void RegisterSyntaxTreeAction(this AnalysisContext context, Action<SyntaxTreeAnalysisContext, StyleCopSettings> action)
```

Comment:
📝 So this is a bit of a strange situation. The `AnalysisContext.TryGetValue<TValue>` method is effectively static - it doesn't access any state of `AnalysisContext`. In theory, you could do something like this:

```csharp
delegate bool TryGetStyleCopSettingsFileJson(SourceText text, SourceTextValueProvider<Lazy<JsonValue>> valueProvider, out Lazy<JsonValue> value);

// This delegate can be captured and used in the callback
TryGetStyleCopSettingsFileJson tryGetValue = context.TryGetValue<Lazy<JsonValue>>;
```

Reply:
For cases where more than one callback is registered, it does seem like it would be more efficient to use an explicit start callback and only compute it once.

Reply:
The big problem right now is re-creating the settings object over and over. Holding on to these two comments for later, when most of the work is done.

Reply:
> For cases where more than one callback is registered, it does seem like it would be more efficient to use an explicit start callback and only compute it once.

Handled in #3648